src/NLog/Targets/FileTarget.cs

```diff
@@ -1884,37 +1881,68 @@ private string GetPotentialFileForArchiving(string fileName)
         /// <returns>Filename to archive. If <c>null</c>, then nothing to archive.</returns>
         private string GetArchiveFileNameBasedOnFileSize(string fileName, int upcomingWriteSize)
         {
-            if (ArchiveAboveSize == ArchiveAboveSizeDisabled)
+            if (ArchiveAboveSize <= ArchiveAboveSizeDisabled)
             {
                 return null;
             }
 
             var previousFileName = GetPotentialFileForArchiving(fileName);
-            if (previousFileName == null)
+            if (string.IsNullOrEmpty(previousFileName))
             {
                 return null;
             }
 
-            var length = _fileAppenderCache.GetFileLength(previousFileName);
-            if (length == null)
+            //this is an expensive call
+            var fileLength = _fileAppenderCache.GetFileLength(previousFileName);
+            if (!fileLength.HasValue)
             {
-                return null;
+                if (!string.IsNullOrEmpty(_previousLogFileName) && previousFileName != _previousLogFileName)
+                {
+                    upcomingWriteSize = 0;
+                    return GetArchiveFileNameBasedOnFileSize(_previousLogFileName, upcomingWriteSize);
+                }
+                else
+                {
+                    return null;
+                }
             }
 
             if (previousFileName != fileName)
             {
                 upcomingWriteSize = 0;  // Not going to write to this file
             }
 
-            var shouldArchive = length.Value + upcomingWriteSize > ArchiveAboveSize;
+            var shouldArchive = ShouldArchiveOnFileSize(fileLength.Value, upcomingWriteSize);
             if (shouldArchive)
             {
-                return previousFileName;
+                _fileAppenderCache.FlushAppenders();    // Flush before double-checking
+
+                FileInfo fileInfo = new FileInfo(previousFileName);
+                if (fileInfo.Exists)
+                {
+                    shouldArchive = ShouldArchiveOnFileSize(fileInfo.Length, upcomingWriteSize);
+                    if (shouldArchive)
+                    {
+                        return previousFileName;
+                    }
+
+                    _fileAppenderCache.InvalidateAppender(previousFileName)?.Dispose(); // File changed behind our backs
```

Comment:
Could we move this out of the if/else? (Same method invocation in if and else branch)

Reply:
Fixed